### DISCUSSION TITLE
Performance when adding many subscriptions at once

### DISCUSSION THREAD

**Discussion by dorbaruchh** [reduxjs/redux-toolkit#2357] [Q&A]
Hello, 
I'm working on a page that includes a table and a filtering column. The table is showing 50 rows. 
Right now when a filter is modified and the table data changes, there is a noticeable lag. 
It seems that the performance issue is coming from a column that renders a component that is subscribing to about 3 different endpoints because when I remove this specific column the filtering performance improves significantly.
The component doesn't have other logic or expensive computation that might explain the performance issue.
I was wondering if removing and then adding relatively many subscriptions at once (at most 50 subscription removed and 50 added) is expected to cause some lag, or is this explanation unlikely and I should look elsewhere?

Thanks!

**Comment by markerikson** [comment]
Can you clarify what you mean by "subscriptions"?

The only "subscriptions" in Redux Toolkit are when components use a `useGetSomeDataQuery` hook from RTK Query, where a "subscription" is a reference counter.

Are you talking about React-Redux and store subscriptions instead?  Or is this indeed RTKQ?

Can you show a CodeSandbox or a repo of what you're trying to do?

**Reply by dorbaruchh** [reply]
Yes, I am talking about useGetSomeDataQuery with a new set of arguments using RTKQ. 
From your answer, I understand that adding subscriptions shouldn't have any impact on performance.

**Reply by markerikson** [reply]
Hmm. It _may_ have an impact on performance, but tbh we haven't tried to measure the cost before.

**Reply by phryneas** [reply]
Generally it's probably the same cost as having a `useSelector` in each of those components.

Have you generally done some measuring or profiling? Is it just too many rerenders, is it too slow component renders, is it inbetween the renders?

**Reply by dorbaruchh** [reply]
From what I can see, it doesn't seem to be too many rerenders. From the moment the filters change and until the rendering stops, the table (and each cell inside of it) only rerenders 3 times.
I tried to replace all subscriptions to endpoints with some default values and the performance improved significantly, and the number of rerenders only decreased by one.
From profiling I can't see any indication for slow component renders.
I disabled redux debugging and the performance also improved, but there is still see a significant difference compared to not using RTKQ.
It might be worth noting that I'm using a 4x CPU slowdown using Chrome developer tools. I'm doing that because I have a pretty strong CPU and can't see performance issues, but other people complained about the performance.

**Reply by markerikson** [reply]
@dorbaruchh how are you measuring "performance", specifically?  Render timings? JS CPU captures? Something else?

**Reply by dorbaruchh** [reply]
I didn't actually measure it, but there is a noticeable difference in the lag you experience after performing the filtering. 
Sorry I can't provide more accurate information, I'm not very experienced in debugging performance issues.

**Reply by markerikson** [reply]
@dorbaruchh can you put together a CodeSandbox or a standalone repo that shows this behavior happening?

**Reply by dorbaruchh** [reply]
I can try, although to fully reproduce this issue I would need to create a relatively elaborate setup. 
Thank you for your help!

**Reply by dorbaruchh** [reply]
@markerikson I put together a simple CodeSandBox:
https://codesandbox.io/s/jolly-night-3ihy1w?file=/src/index.tsx

Although pretty simple, I think this example is still quite similar to the behavior I have in my app.
Each LiveComponent is subscribing to an endpoint that updates the data every second (in my app the data is updated with a websocket). When you click the button on the top of the page the arguments for all the components change which causes them to unsubscribe and then resubscribe with the new arguments (this basically simulates the filtering in my app). 
With 4x CPU slowdown I think you'll be able to see the lag I mentioned.

**Reply by dorbaruchh** [reply]
@markerikson @phryneas I was wondering if you got a chance to look at the CodeSandbox I created?
I continued to do testing and it was pretty clear that subscribing from many components had a pretty significant impact on performance. I'm trying to understand if I'm doing something wrong or if this is expected.
For now I had to remove usages of RTKQ in all components that might have many instances rendered on the screen at once, for example a component that is rendered inside a big table. I ended up replacing some of the direct subscriptions from the inner smaller components with a single subscription from the parent and passed the information down to the children using React context.
I also migrated my websocket implementation to stop using RTKQ because I need to be able to subscribe to websockets from many small components at once like in the example I provided.

**Reply by markerikson** [reply]
@dorbaruchh Sorry, no - haven't had much time to do much of anything "maintainer"-ish lately, and tbh this issue is nowhere on my radar right now.

If you could maybe make a JS perf recording with the Chrome DevTools, zip the JSON, and attach it here, I might be able to take a look and get a sense of where the bottlenecks are.

**Reply by phryneas** [reply]
Unfortunately, I also don't have too much time, but I can give you a hint:
If you have components, where you know that another mounted component definitely also uses the same query, you can do a `api.endpoints.foobar.useQueryState` hook call instead of the `useFoobarQuery` call. That one really just picks the data out of the store and does not create a subscription (but will also not cause data to be fetched or be held in cache as a consequence).

**Reply by markerikson** [reply]
@dorbaruchh I've just reworked the internals of RTKQ to minimize the number of actions dispatched when setting up subscriptions - see #2599 .

Could you try out the PR Preview build and see if that works better for you?

**Reply by zkbule** [reply]
I have similar issue. Migrated from Thunk to RTKQ but see lag afterwards. I think it's related to the internal subscription update because when the component mount/unmounted (switch tabs), there's a **_[Violation] 'click' handler took ***s_** message in the console. I can confirm that there's no complicated data select and re-render.

![image](https://user-images.githubusercontent.com/13895180/220464607-43b342d7-2aeb-47e1-bc4f-9ef2b1d62cf9.png)

**Reply by markerikson** [reply]
@zkbule : could you post a CodeSandbox or a Github repo that shows this happening?

Also, what version of Redux Toolkit are you using specifically?  We made a lot of perf improvements to this in RTK 1.9.

**Reply by zkbule** [reply]
> @zkbule : could you post a CodeSandbox or a Github repo that shows this happening?
> 
> Also, what version of Redux Toolkit are you using specifically? We made a lot of perf improvements to this in RTK 1.9.

I'm using 1.9.3. I'll find some time to create a CodeSandbox

**Reply by zkbule** [reply]
> > @zkbule : could you post a CodeSandbox or a Github repo that shows this happening?
> > Also, what version of Redux Toolkit are you using specifically? We made a lot of perf improvements to this in RTK 1.9.
> 
> I'm using 1.9.3. I'll find some time to create a CodeSandbox

Here's the sandbox:
https://codesandbox.io/s/damp-morning-v6108m

![image](https://user-images.githubusercontent.com/13895180/220495120-09e9ffb9-7c2d-40bc-8acd-4444219e11f3.png)


I think it's related to the listenerMiddleware (In my real app, it's more obvious if I disable the listener middleware)

**Reply by markerikson** [reply]
Hmm. @zkbule : I tried opening up the sandbox app in a separate tab and did a perf recording while clicking back and forth, and I don't see any warnings about event handlers taking a long time.  

Looking at the JS perf capture, the biggest things taking up time in the sandbox are CSB's own console processing, and a couple spots where RTK's dev-mode checks run during dispatch.

Does this still happen in your app if you try turning off the RTK serializable check middleware?

**Reply by zkbule** [reply]
> Hmm. @zkbule : I tried opening up the sandbox app in a separate tab and did a perf recording while clicking back and forth, and I don't see any warnings about event handlers taking a long time.
> 
> Looking at the JS perf capture, the biggest things taking up time in the sandbox are CSB's own console processing, and a couple spots where RTK's dev-mode checks run during dispatch.
> 
> Does this still happen in your app if you try turning off the RTK serializable check middleware?

I turned off the serialization check middleware but it didn’t help much. However, I noticed that one of my three listeners had some large data set which caused the lag when it was on (inside effect). For now, it’s not a blocker for me to continue using RTKQ since I can use another strategy to eliminate that listener. (Though I’m still not sure why the thunk + reselector approach was not impacted). Anyway, thank you for looking into this.

**Reply by markerikson** [reply]
@zkbule out of curiosity, what does that listener code look like?  What data set are you working with, and what are you doing with it?

**Reply by zkbule** [reply]
> @zkbule out of curiosity, what does that listener code look like? What data set are you working with, and what are you doing with it?

It's for telemetry. Basically, listen to all actions and log (Application Insights track event) store states.

` startAppListening({
      predicate: action => {
        // subscribe all actions
        return typeof action.type === "string" && action.type !== "";
      },
      effect: (action, { getState }) => {
        const properties = JSON.stringify(getState());
        trackEvent(action.type, { state: JSON.parse(properties) });
      }
    });`

**Reply by markerikson** [reply]
@zkbule oh my. Yeah, if you're doing a `JSON.stringify()` of the _entire_ state _every_ time, and your state has any real size or depth to it, that's going to be expensive. Please _don't_ do that!

Is there a reason why you feel you need to make a full clone of the Redux state for _every_ action?  

At most, I'd suggest passing along the current action contents, but _not_ the whole state.

**Reply by zkbule** [reply]
> @zkbule oh my. Yeah, if you're doing a `JSON.stringify()` of the _entire_ state _every_ time, and your state has any real size or depth to it, that's going to be expensive. Please _don't_ do that!
> 
> Is there a reason why you feel you need to make a full clone of the Redux state for _every_ action?
> 
> At most, I'd suggest passing along the current action contents, but _not_ the whole state.

Yea...I realize that. Actually the most time consuming part is the trackEvent, not the json stringfy. And you are right, I'm logging action payload now instead of store state. Do you have any idea why it's not impact the thunk + reselect approach?

**Reply by markerikson** [reply]
I would expect this to impact thunks _and_ normal actions as well, because it was running on _every_ dispatched action.

**Reply by zkbule** [reply]
> I would expect this to impact thunks _and_ normal actions as well, because it was running on _every_ dispatched action.

I wish I can show you my thunk version with the listener. One question, isn't the effect in listener supposed to be async? So the expensive operation in the effect shouldn't impact the overall action dispatch? (I know it's bad...but just out of curiosity)

**Reply by markerikson** [reply]
@zkbule : no, listeners run synchronously as part of the overall dispatch sequence, just like any other code in a middleware.  Even if you mark the effect function as `async`, all the code up until the first `await` is still synchronous:

```ts
startAppListening({
  actionCreator: todoAdded,
  effect: async (action, listenerApi) => {
    console.log("This runs sync as part of the dispatch");
    await listenerApi.delay(1);
    console.log("This runs in a new event loop tick");
  }
})
```

(Of course, any expensive code that takes a long time to run is still going to block the event loop for that long, no matter whether it's in a click handler or a `setTimeout` or after an `await` - it's just a question of _when_ that expensive code runs.)

And yeah, if you can show a more realistic example of what you've got in your app atm, that would be helpful, especially if it's a sandbox I can run and look at.
